bin/refresh_matviews.pl

```diff
@@ -13,89 +13,115 @@ =head1 DESCRIPTION
 
 =head1 AUTHOR
 
 Bryan Ellerbrock <bje24@cornell.edu>
+Naama Menda <nm249@cornell.edu>
 
 =cut
 
 use strict;
 use warnings;
-use Getopt::Std;
 use DBI;
-#use CXGN::DB::InsertDBH;
+use Try::Tiny;
+use Getopt::Long;
 
-our ($opt_H, $opt_D, $opt_U, $opt_P, $opt_m, $opt_c, $refresh, $status);
-getopts('H:D:U:P:m:c');
+my ( $dbhost, $dbname, $username, $password, $mode, $concurrent, $test);
+GetOptions(
+    'm=s'        => \$mode,
+    'c'          => \$concurrent,
+    'P=s'        => \$password,
+    'U=s'        => \$username,
+    't'          => \$test,
+    'dbname|D=s' => \$dbname,
+    'dbhost|H=s' => \$dbhost,
+);
+
+
+unless ($mode =~ m/^(fullview|stockprop|phenotypes)$/ ) { die "Option -m must be fullview, stockprop, or phenotypes. -m  = $mode\n"; }
 
 print STDERR "Connecting to database...\n";
-my $dsn = 'dbi:Pg:database='.$opt_D.";host=".$opt_H.";port=5432";
-my $dbh = DBI->connect($dsn, $opt_U, $opt_P);
+my $dsn = 'dbi:Pg:database='.$dbname.";host=".$dbhost.";port=5432";
+my $dbh = DBI->connect($dsn, $username, $password, { RaiseError => 1, AutoCommit=>0 });
 
-eval {
+try {
     print STDERR "Refreshing materialized views . . ." . localtime() . "\n";
-
-    if ($opt_m eq 'fullview'){
-        my $q = "UPDATE public.matviews SET currently_refreshing=?";
-        my $state = 'TRUE';
-        my $h = $dbh->prepare($q);
-        $h->execute($state);
-
-        if ($opt_c) {
-            $refresh = 'SELECT refresh_materialized_views_concurrently()';
-        } else {
-            $refresh = 'SELECT refresh_materialized_views()';
-        }
-
-        $h = $dbh->prepare($refresh);
-        $status = $h->execute();
-
-        if ($opt_c) {
-          $refresh = 'SELECT refresh_materialized_stockprop_concurrently()';
-        } else {
-          $refresh = 'SELECT refresh_materialized_stockprop()';
-        }
-
-        $h = $dbh->prepare($refresh);
-        $status = $h->execute();
-
-        if ($opt_c) {
-          $refresh = 'SELECT refresh_materialized_phenotype_jsonb_table_concurrently()';
-        } else {
-          $refresh = 'SELECT refresh_materialized_phenotype_jsonb_table()';
-        }
-
-        $h = $dbh->prepare($refresh);
-        $status = $h->execute();
-
-        $q = "UPDATE public.matviews SET currently_refreshing=?";
-        $state = 'FALSE';
-        $h = $dbh->prepare($q);
-        $h->execute($state);
+    my %matviews; 
+    my $matviews_q = "SELECT mv_name, mv_dependents FROM matviews";
+    my $matviews_h = $dbh->prepare($matviews_q);
+    $matviews_h->execute();
+    while (my ($mv_name, $mv_dependents) = $matviews_h->fetchrow_array() ) {
+	$matviews{$mv_name} = $mv_dependents ;
+    }
+    
+    my $cur_refreshing_q =  "UPDATE public.matviews SET currently_refreshing=?";
+    my $state = 'TRUE';
+    
+    my @mv_names = ();
+    
+    if ($mode eq 'fullview') {
+	@mv_names = keys ( %matviews );
+	push (@mv_names, "materialized_stockprop", "materialized_phenotype_jsonb_table");
```

Comment:
Hey @nmenda , with the db patch update to the matviews table this line should no longer be needed, right?

Also I'm thinking we should add 'phenotype' and 'genotype' modes options here. But I see what you mean about needing many to many relationships to do it, maybe via a new matview_types table

Reply:
I removed this line and added a 'refresh_start' column to matviews. 
Will add more modes as we decide how to group the views in the db